bullet-featherstone/src/SDFFeatures.cc

```diff
@@ -1051,34 +1051,92 @@ bool SDFFeatures::AddSdfCollision(
+            centroid *= 1.0/static_cast<double>(submesh.VertexCount());
+            btAlignedObjectArray<btVector3> vertices;
+            for (std::size_t i = 0; i < submesh.VertexCount(); ++i)
+            {
+              gz::math::Vector3d v = submesh.Vertex(i) - centroid;
+              vertices.push_back(convertVec(v) * scale);
+            }
+
+            float collisionMargin = 0.001f;
+            this->meshesConvex.push_back(std::make_unique<btConvexHullShape>(
+                &(vertices[0].getX()), vertices.size()));
+            auto *convexShape = this->meshesConvex.back().get();
+            convexShape->setMargin(collisionMargin);
+
+            btTransform trans;
+            trans.setIdentity();
+            trans.setOrigin(convertVec(centroid) * scale);
+            compoundShape->addChildShape(trans, convexShape);
+          }
+        }
+        meshCreated = true;
       }
 
-      this->triangleMeshes.push_back(std::make_unique<btTriangleMesh>());
-      for (unsigned int i = 0; i < indexCount/3; i++)
+      if (!meshCreated)
       {
-        const btVector3& v0 = convertedVerts[s->Index(i*3)];
-        const btVector3& v1 = convertedVerts[s->Index(i*3 + 1)];
-        const btVector3& v2 = convertedVerts[s->Index(i*3 + 2)];
-        this->triangleMeshes.back()->addTriangle(v0, v1, v2);
-      }
+        auto vertexCount = s->VertexCount();
+        auto indexCount = s->IndexCount();
+        btAlignedObjectArray<btVector3> convertedVerts;
+        convertedVerts.reserve(static_cast<int>(vertexCount));
+        for (unsigned int i = 0; i < vertexCount; i++)
+        {
+          convertedVerts.push_back(btVector3(
+                static_cast<btScalar>(s->Vertex(i).X()) * scale[0],
+                static_cast<btScalar>(s->Vertex(i).Y()) * scale[1],
+                static_cast<btScalar>(s->Vertex(i).Z()) * scale[2]));
+        }
 
-      this->meshesGImpact.push_back(
-        std::make_unique<btGImpactMeshShape>(
-          this->triangleMeshes.back().get()));
-      this->meshesGImpact.back()->updateBound();
-      this->meshesGImpact.back()->setMargin(btScalar(0.01));
-      compoundShape->addChildShape(btTransform::getIdentity(),
-        this->meshesGImpact.back().get());
+        this->triangleMeshes.push_back(std::make_unique<btTriangleMesh>());
+        for (unsigned int i = 0; i < indexCount/3; i++)
+        {
+          const btVector3& v0 = convertedVerts[s->Index(i*3)];
+          const btVector3& v1 = convertedVerts[s->Index(i*3 + 1)];
+          const btVector3& v2 = convertedVerts[s->Index(i*3 + 2)];
+          this->triangleMeshes.back()->addTriangle(v0, v1, v2);
+        }
+
+        this->meshesGImpact.push_back(
+          std::make_unique<btGImpactMeshShape>(
+            this->triangleMeshes.back().get()));
+        this->meshesGImpact.back()->updateBound();
+        this->meshesGImpact.back()->setMargin(btScalar(0.01));
+        compoundShape->addChildShape(btTransform::getIdentity(),
```

Comment:
this currently uses `btCompoundShape` to join submeshes

Reply:
Updated to merge all submeses before decomposition. 

Note that the decomposed submeshes are then still joined together using `btCompoundShape`. Same for the unoptimized case. I can change this to using a single mesh in a follow-up PR.